sw/device/lib/ujson/ujson_derive.h

```diff
@@ -0,0 +1,264 @@
+    status_t ujson_serialize_##name_(ujson_t *uj, const name_ *self) { \
+        switch(*self) { \
+            decl_(formal_name_, ujson_ser_enum) \
+            default: \
+                TRY(ujson_putbuf(uj, \
+                    "{\"" RUST_ENUM_INTVALUE_STR "\":", \
+                    sizeof("{\"" RUST_ENUM_INTVALUE_STR "\":") - 1)); \
+                TRY(ujson_serialize_uint32_t(uj, (const uint32_t*)self)); \
+                TRY(ujson_putbuf(uj, "}", 1)); \
+        } \
+        return OK_STATUS(); \
+    } \
+    extern const int __never_referenced___here_to_eat_a_semicolon[]
+
+//////////////////////////////////////////////////////////////////////
+// Deserialize Implementation
+//////////////////////////////////////////////////////////////////////
+#define ujson_de_loop_indirect() ujson_de_loop
+#define ujson_de_loop(mult, expr, count, ...) \
+    TRY(ujson_consume(uj, '[')); \
+    OT_IIF(OT_NOT(OT_VA_ARGS_COUNT(dummy, ##__VA_ARGS__))) \
+    ( /*then*/ \
+        size_t i = 0; \
+        while(true) { \
+            if (TRY(ujson_consume_maybe(uj, ']'))) break; \
+            if (i) TRY(ujson_consume(uj, ',')); \
+            if (i < count) {  expr; ++i; } \
+        } \
+        if (i < count) { p += (count - i) * mult; } \
+    , /*else*/ \
+        for(size_t x=0;; ++x) { \
+            if (TRY(ujson_consume_maybe(uj, ']'))) break; \
+            if (x) TRY(ujson_consume(uj, ',')); \
+            OT_OBSTRUCT(ujson_de_loop_indirect)()(mult, expr, __VA_ARGS__) \
+        } \
+    ) /*endif*/ \
+
+#define ujson_de_field(name_, type_, ...) \
+    else if (ujson_streq(key, #name_)) { \
+        OT_IIF(OT_NOT(OT_VA_ARGS_COUNT(dummy, ##__VA_ARGS__))) \
+        ( /*then*/ \
+            TRY(ujson_deserialize_##type_(uj, &self->name_)); \
+        , /*else*/ \
+            type_ *p = (type_*)self->name_; \
+            OT_EVAL(ujson_de_loop(1, \
+                TRY(ujson_deserialize_##type_(uj, p++)), __VA_ARGS__)) \
+        ) /*endif*/ \
+    }
+
+#define ujson_de_string(name_, size_, ...) \
+    else if (ujson_streq(key, #name_)) { \
+        OT_IIF(OT_NOT(OT_VA_ARGS_COUNT(dummy, ##__VA_ARGS__))) \
+        ( /*then*/ \
+            TRY(ujson_parse_qs(uj, self->name_, sizeof(self->name_))); \
+        , /*else*/ \
+            char *p = (char*)self->name_; \
+            OT_EVAL(ujson_de_loop(size_, \
+                TRY(ujson_parse_qs(uj, p, sizeof(self->name_))); p+=size_, __VA_ARGS__)) \
+        ) /*endif*/ \
+    }
+
+#define UJSON_DESERIALIZE_STRUCT(name_, decl_) \
+    status_t ujson_deserialize_##name_(ujson_t *uj, name_ *self) { \
+        size_t nfield = 0; \
+        char key[128]; \
+        TRY(ujson_consume(uj, '{')); \
+        while(TRY(ujson_consume_maybe(uj, '}')) == 0) { \
+            if (nfield++ > 0) { \
+                TRY(ujson_consume(uj, ',')); \
+            } \
+            TRY(ujson_parse_qs(uj, key, sizeof(key))); \
+            TRY(ujson_consume(uj, ':')); \
+            if (0) {} \
```

Comment:
Out of curiosity, I see this a lot, what is the purpose of the `if (0) {}`? Is it strictly to accommodate the `decl_()` clause being an `else if ...`?

Reply:
That's exactly what its for:  The decl_() is going to emit an `else if (cond) { stuff; }` for each struct member.